python/cudf/dataframe/multiindex.py

```diff
@@ -200,25 +242,40 @@ def _get_column_major(self, df, row_tuple):
         return result
 
     def __len__(self):
-        return len(self.codes[self.codes.columns[0]])
+        if hasattr(self, '_source_data'):
+            return len(self._source_data)
+        else:
+            return len(self.codes[self.codes.columns[0]])
 
     def __eq__(self, other):
-        if not hasattr(other, 'levels'):
+        if not hasattr(other, '_levels'):
             return False
-        equal_levels = self.levels == other.levels
-        if isinstance(equal_levels, np.ndarray):
-            equal_levels = equal_levels.all()
-        return equal_levels and\
-            self.codes == other.codes and\
-            self.names == other.names
+        # Lazy comparison
+        if hasattr(other, '_source_data'):
+            for col in self._source_data.columns:
+                if col not in other._source_data.columns:
+                    return False
+                if not self._source_data[col].equals(other._source_data[col]):
+                    return False
+            return True
+        else:
+            # Lazy comparison isn't possible - MI was created manually.
+            # Actually compare the MI, not its source data (it doesn't have
+            # any).
+            equal_levels = self.levels == other.levels
+            if isinstance(equal_levels, np.ndarray):
+                equal_levels = equal_levels.all()
+            return equal_levels and\
+                self.codes == other.codes and\
+                self.names == other.names
 
     @property
     def is_contiguous(self):
         return True
 
     @property
     def size(self):
-        return len(self.codes[0])
+        return len(self._source_data)
```

Comment:
You check if `_source_data` exists in other functions before using it. We either need to make it mandatory or check if it exists here before asking for the length.

Reply:
👍